### ISSUE TITLE
Batched Text Generation Causes Degraded Results

### ISSUE THREAD

**Issue by JC-LMCO** [linkedin/Liger-Kernel#544]
### 🐛 Describe the bug

I'm fine-tuning a llama model using and I've noticed that when I'm generating text with my fine-tuned model, the results are different (and worse) if I am batching my inputs versus when I generate with a single sequence (even if the tokens, including padding, are exactly the same). Now, generating using batching is known to cause different outputs (see huggingface/transformers#23017), however this shouldn't be the case when I'm not sampling and using FP32 weights. When I use `AutoModelForCausalLM`, the tokens and logits are very similar with batch and non-batch generation. However, when I use `AutoLigerKernelForCausalLM`, the logits are significantly different when doing batched vs non-batched generation. 

I've included the code that I use to test batched vs non-batched below and HF vs LigerKernel models on the exact same inputs.

### Reproduce

```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from liger_kernel.transformers import AutoLigerKernelForCausalLM

# Load tokenizer
tokenizer = AutoTokenizer.from_pretrained('problematic_model')

# Load batch
batch = torch.load('error_inputs.pt')

BAD_SEQ_IDX = 2

# Non-batched
batch_non_batched = {
    'input_ids': batch['input_ids'][BAD_SEQ_IDX].unsqueeze(0).cuda(),
    'attention_mask': batch['attention_mask'][BAD_SEQ_IDX].unsqueeze(0).cuda()
}

# Non-batched with AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained('problematic_model', torch_dtype=torch.bfloat16)
model.cuda()

gen_outs_non_batched = model.generate(**batch_non_batched, max_new_tokens=200, output_logits=True, return_dict_in_generate=True)
logits_sort, tokens_sort = gen_outs_non_batched.logits[3][0].sort(descending=True)

print("Non-batched with AutoModelForCausalLM:")
print("Top 5 tokens:", tokens_sort[:5].tolist())
print('Top 5 logits:', logits_sort[:5].tolist())
print("Sequence:", tokenizer.decode(gen_outs_non_batched.sequences[0], skip_special_tokens=True))

# Batched with AutoModelForCausalLM
gen_outs_batched = model.generate(**batch, max_new_tokens=100, output_logits=True, return_dict_in_generate=True)
logits_sort, tokens_sort = gen_outs_batched.logits[3][BAD_SEQ_IDX].sort(descending=True)

print("\n\nBatched with AutoModelForCausalLM:")
print("Top 5 tokens:", tokens_sort[:5].tolist())
print('Top 5 logits:', logits_sort[:5].tolist())
print("Sequence:", tokenizer.decode(gen_outs_batched.sequences[BAD_SEQ_IDX], skip_special_tokens=True))

# Non-batched with AutoLigerKernelForCausalLM
model = AutoLigerKernelForCausalLM.from_pretrained('problematic_model', torch_dtype=torch.bfloat16, attn_implementation="sdpa")
model.cuda()

gen_outs_non_batched_liger = model.generate(**batch_non_batched, max_new_tokens=200, output_logits=True, return_dict_in_generate=True)
logits_sort, tokens_sort = gen_outs_non_batched_liger.logits[3][0].sort(descending=True)

print("\n\nNon-batched with AutoLigerKernelForCausalLM:")
print("Top 5 tokens:", tokens_sort[:5].tolist())
print('Top 5 logits:', logits_sort[:5].tolist())
print("Sequence:", tokenizer.decode(gen_outs_non_batched_liger.sequences[0], skip_special_tokens=True))

# Batched with AutoLigerKernelForCausalLM
gen_outs_batched_liger = model.generate(**batch, max_new_tokens=100, output_logits=True, return_dict_in_generate=True, do_sample=False)
logits_sort, tokens_sort = gen_outs_batched_liger.logits[3][BAD_SEQ_IDX].sort(descending=True)

print("\n\nBatched with AutoLigerKernelForCausalLM:")
print("Top 5 tokens:", tokens_sort[:5].tolist())
print('Top 5 logits:', logits_sort[:5].tolist())
print("Sequence:", tokenizer.decode(gen_outs_batched_liger.sequences[BAD_SEQ_IDX], skip_special_tokens=True))


# Batched with AutoLigerKernelForCausalLM in FP32
model = AutoLigerKernelForCausalLM.from_pretrained('problematic_model', torch_dtype=torch.float32, attn_implementation="sdpa")
model.cuda()

gen_outs_batched_liger = model.generate(**batch, max_new_tokens=100, output_logits=True, return_dict_in_generate=True, do_sample=False)
logits_sort, tokens_sort = gen_outs_batched_liger.logits[3][BAD_SEQ_IDX].sort(descending=True)

print("\n\nBatched with FP32 AutoLigerKernelForCausalLM:")
print("Top 5 tokens:", tokens_sort[:5].tolist())
print('Top 5 logits:', logits_sort[:5].tolist())
print("Sequence:", tokenizer.decode(gen_outs_batched_liger.sequences[BAD_SEQ_IDX], skip_special_tokens=True))
```
Prints

> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [22.0, 16.375, 14.625, 14.5625, 14.3125]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with AutoModelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [22.0, 16.375, 14.625, 14.5, 14.3125]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Non-batched with AutoLigerKernelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [22.0, 16.375, 14.625, 14.5, 14.3125]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with AutoLigerKernelForCausalLM:
> Top 5 tokens: [28, 260, 338, 365, 43]
> Top 5 logits: [15.375, 13.1875, 12.875, 12.6875, 12.4375]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt, "The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with FP32 AutoLigerKernelForCausalLM:
> Top 5 tokens: [28, 260, 338, 365, 43]
> Top 5 logits: [15.20411491394043, 12.999839782714844, 12.732202529907227, 12.52097225189209, 12.336112976074219]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt, "The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--"The truth's a-oyed,--"--

### Versions

Environment Report:
-------------------
Operating System: Linux-5.14.0-427.47.1.el9_4.x86_64-x86_64-with-glibc2.31
Python version: 3.10.13
Liger Kernel version: 0.5.2
PyTorch version: 2.5.1+cu121
CUDA version: 12.1
HIP(ROCm) version: Not available
Triton version: 3.1.0
Transformers version: 4.48.1
XPU version: XPU Not Available

**Comment by ByronHsu** [comment]
@Tcc0403  can you help take a look? @JC-LMCO may you provide which llama model you use? this seems like a numerical issue to me

**Comment by JC-LMCO** [comment]
I was fine-tuning [HuggingFaceTB/SmolLM2-135M](https://huggingface.co/HuggingFaceTB/SmolLM2-135M). This model is about .1 epochs into training. Since it's a base model, the output is gibberish with or without batching, however when I re-ran my script with `HuggingFaceTB/SmolLM2-135M`, I observed the same results where there is a significant difference b/w the logits with batching vs w/o batching for Liger (and again, AutoModel had very similar values regardless of batch size).

**Comment by Tcc0403** [comment]
Could you re-run your script with liger-kernel-nightly? My guess is the batch bug in rope fixed by https://github.com/linkedin/Liger-Kernel/commit/d7c78df1455cac50205ad4a9b9ac50c5d144ba0f

**Comment by JC-LMCO** [comment]
Okay there we go, much better! 

> Non-batched with AutoModelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [22.0, 16.375, 14.625, 14.5625, 14.3125]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with AutoModelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [22.0, 16.375, 14.625, 14.5, 14.3125]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Non-batched with AutoLigerKernelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [22.0, 16.375, 14.625, 14.5, 14.3125]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with AutoLigerKernelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [21.875, 16.25, 14.5625, 14.5, 14.1875]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with FP32 AutoLigerKernelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [21.829843521118164, 16.17617416381836, 14.49455451965332, 14.355961799621582, 14.183671951293945]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.

There is still a minor discrepancy b/w batched and non-batched that doesn't appear in the standard HuggingFace implementation and the non-batched version matches the HF implementation.

**Comment by JC-LMCO** [comment]
While the logits are significantly closer, I've noticed the batch size still causes a discrepancy in the generation that isn't present in the HF implementation. Do you think that this is just a numerical instability thing, or some other subtle problem?

**Comment by Tcc0403** [comment]
Thanks for your test and response. The discrepancy between batched and non-batched in Liger Implementation might be a numerical issue. 

There are some unavoidable upcastings in Liger, so Liger and HF can't be a perfect match in bf16. Could you run your script to compare just HF impl fp32 and Liger impl fp32? If logits still differ that much between liger fp32 and hf fp32, then it might be other problems that we missed. I'll look into it if the discrepancy still remain.

**Comment by JC-LMCO** [comment]
With everything in FP32, looks like the logits all match up (to the acceptable level of precision). I guess the difference b/w HF and Liger at Fp16 is just down to floating point errors propagating in different ways, and that the behavior I was seeing earlier was just the rope scaling problem. Thank you!

> Non-batched with AutoModelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [21.829837799072266, 16.1761531829834, 14.494553565979004, 14.355947494506836, 14.183673858642578]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with AutoModelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [21.82984733581543, 16.176164627075195, 14.494560241699219, 14.35595989227295, 14.183674812316895]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Non-batched with AutoLigerKernelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [21.8298397064209, 16.1761531829834, 14.49455451965332, 14.355948448181152, 14.183676719665527]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.
> 
> 
> Batched with AutoLigerKernelForCausalLM:
> Top 5 tokens: [260, 28, 476, 365, 378]
> Top 5 logits: [21.829843521118164, 16.17617416381836, 14.49455451965332, 14.355961799621582, 14.183671951293945]
> Sequence: INPUT:
> they learnt the contrary afterwards but at this time the words looked like a decoy that had no reality merely intended to attract and catch the huguenots on palm sunday a very learned dominican as it was said come from auxerre was commissioned to preach the victory of rome
> 
> OUTPUT:
> They learnt the contrary afterwards, but at this time the words looked like a decoy that had no reality, merely intended to attract and catch the Huguenots. On Palm Sunday a very learned Dominican, as it was said, "came from Auxerre," was commissioned to preach the victory of Rome.
> Setting `pad_token_id` to `eos_token_id`:0 for open-end generation.

**Comment by JC-LMCO** [comment]
I'll go ahead and close this out b/c after pulling from nightly and getting the change to rope that got fixed in #477 there doesn't seem to be anything but 16-bit numerical errors causing a difference b/c HF and LigerKernel.
